### ISSUE TITLE
Expectation protocol

### ISSUE THREAD

**Issue by kevinsung** [quantumlib/Cirq#1901]
PauliSum, LinearCombinationOfGates, and Circuit could all implement a protocol
```
def _expectation_(self, state: Union[int, np.ndarray]) -> complex
```

**Comment by peterse** [comment]
Do you have in mind what the function that calls this protocol look like? 

I was thinking of a single `expectation` protocol that would dispatch to different `_expectation_*`  methods defined for wavefunctions, density matrices, or bitstring arrays. Then, for example,  `expectation(cirq.X(cirq.LineQubit(0)), wf)`, would end up using a method like `PauliString._expectation_from_wavefunction`.

Right now, if we were to put this method into `PauliString` it would just be a thin wrapper for `cirq.pauli_string_expectation(self).value_derived_from_wavefunction`. However, it feels like the methods of `display.ApproxPauliStringExpectation` and `display.PauliStringExpectation` should live inside `PauliString`.

**Comment by peterse** [comment]
@Strilanc @mpharrigan does this seem like a reasonable protocol?

**Comment by mpharrigan** [comment]
xref #1805 

@peterse : this seems like a weird form of dispatch. Why not `wf.expectation()`? 

@kevinsung : What is state? How would the protocol work on `Circuit`s? Circuits are unitary but generally not Hermitian

**Comment by peterse** [comment]
@mpharrigan I don't think there's a dedicated class for wavefunctions/dm's at this point.

Another alternative is `PauliString.expectation( ... )` though the natural choice of defining `PaulliSum.expectation` in terms of `PauliString.expectation` would lead to a lot of redundant input validating/dispatch.

**Comment by kevinsung** [comment]
@mpharrigan `state` would be the wavefunction. Now that you mention it, in fact none of the objects I mentioned (PauliSum, LinearCombinationOfGates, and Circuit) are guaranteed to be Hermitian. I was thinking that the expectation for any object A would nevertheless be defined as ⟨Ѱ|A|Ѱ〉, but I guess this might not be conventional?

**Comment by peterse** [comment]
resolved with #1975
